### PR TITLE
Add more types of interaction

### PR DESCRIPTION
**Pull request by hansl**
These types are useful for the project I'm working on, which assumes that left/right are page up/down. Using "next" 10 times does not really help in this case (I want to clamp selection).

I also added a simple Clamped interaction type which does clamp selection and prevent going to the end of the menu from the beginning.

I could not find a simpler pattern here than asking the interaction controller where the selection should go with regard to the interaction type. I tried doing it in my own code by having a custom interaction controller but it ended up being too complex and requiring more logic than simply adding this method to interaction controller here.

I also added a default implementation to prevent making this a breaking change. Technically only new features were added and existing code should not need to be modified.

### CODE REVIEW

**Comment by bugadani** [comment]
I'm a bit conflicted but that may be just that I'm not fully awake yet. Sorry if I'm incoherent at places.

I like the idea to introduce more interaction types. My use cases never needed anything more than "next" and "select", so I never thought about more. I added the "programmed" interactions mostly just to support keyboard arrows in tests. More here is definitely better, however!

It wouldn't be unreasonable to add a direct jump to a particular index, either. You don't have to do this, but I should remember to :)

I got the naming completely wrong. I called them `InteractionController`s, but in truth they are supposed to just translate the input from whatever concept to what embedded-menu can understand. I intended them to, e.g, map keyboard events to InteractionType variants, but not to actually calculate the selection changes.
I'm not sure yet completely, but I think it would be nice to keep this distinction, and just have a single implementation of InteractionType -> selection change.

So why I'm conflicted:
I like that what you did just clicks together, I like the aspect of configurability. I'm just not sure we should do it here. First, I don't think it unreasonable to just implement the clamping behavior. Second, the clamping can be nested. This isn't a problem, but this points to a somewhat clunky design that is not entirely your doing.

I think instead of a wrapper type and overrideable `select`, we might instead create different interactions: e.g. `ForwardWrapping` that would just overflow and `Forward` that would stop at the end first.

Long story short:
 - I think `Forward` and `Backward` can just clamp by default. Pressing page down and sometimes going up in the list would just be weird IMO.
 - `InteractionController`s shouldn't be able to override `select`. We just shouldn't need to have this configurable, or, we should redesign how we handle input.
 - We could have more interactions like `ForwardWrapping` or `JumpTo`, but we don't necessarily need them, or need to add them in this PR.

The examples are getting a bit out of hand :) I admit that I just added nonsense, but some of that nonsense is actually a bad example. For example, by default the detail strings aren't displayed, but I forgot to remove them (or, to show how to enable them) from the examples where this is the case. Also, at this point, we can probably just create a `SdlKeyboardInteraction` type, and add to the library behind a feature flag. We use it in so many places that it just makes sense.

Thank you for working on embedded-menu!

**Comment by hansl** [comment]
So a few things;

Keep in mind I was trying to do all this in a backward compatible way. I didn't want to change existing types, only add new things.

The code in `menu.rs` handling the selection should be refactored out.

> I got the naming completely wrong

I'll make a change to `InputAdapter` which is probably what you had in mind. But why does it also have `fill_area_width`?

> instead create different interactions

I think interactions should be configurable and composable. That would definitely make the different pieces more reusable all around. I don't know about a "ForwardWrapped", etc. I'd rather have a `with_selector(Wrapped)` on the Menu itself. I'll look more into that tonight.

> The examples are getting a bit out of hand

Some of them could be doc examples. There should be more tests as well, and in general more documentation ;-) I'm not going to fault you for that though.

> Thank you for working on embedded-menu!

Spirit of open source. I appreciate the project. It's all good.

----

TL:DR; Let me try to redo this tonight with this comment more in mind.

**Comment by bugadani** [comment]
> Keep in mind I was trying to do all this in a backward compatible way

Feel free to break anything, though I probably don't want to relearn my own library from scratch :) As you can see from the code, this has always been quite low effort, mainly for my own pet project. We can try to turn it into something decent, though I don't think it's worth investing tons of time into if you don't absolutely need to.

> I'll make a change to InputAdapter 

Sounds good!

> But why does it also have fill_area_width?

It probably should have been called `input_progress` or something, it's mainly for the single touch stuff. That is, keep something touched/pressed until it activates, this was the way I added some feedback. You can leave this part to me, if you don't want to deal with it.

> I think interactions should be configurable and composable.

We might have differing opinions here, but only because I can't imagine why would someone want different things to happen when they input a "next" event :) . I'll look forward to what you come up with!

**Comment by bugadani** [comment]
I took the liberty of resolving the `InputAdapter` question, and I'm actually pretty okay with the result. I was able to remove some of the code from `Menu::update` that has long been bugging me, so thanks for the idea!

This had some collateral changes, e.g. from now on the selection indicators calculate the fill_width but they don't yet support actually drawing their own input progress indication. This is good enough for now, but in the future, we might want to indicate progress with a color transition for example where this won't be good enough.

Since you're working on interaction types, I wonder if you have an opinion on how we could support "action on button up" semantics. I'd like to be able to display if a button has been pressed (as `InputState::InProgress(max)` maybe, with some debouncing to avoid flashing), and fire the input event on release.

My little partisan-action unfortunately caused quite a bit of conflicts for this PR. Sorry about that.

**Comment by hansl** [comment]
I'm not going to resolve those conflicts, and will only make the new interaction types in a separate PR.
